### ISSUE TITLE
Unable to connect my Zwift Play to QZ

### ISSUE THREAD

**Issue by felixlbr** [cagnulein/qdomyos-zwift#2458]
I followed the tutorial on Youtube to use my controllers Zwift Play on QZ but it doesn’t work. Even my Home Trainer that I am able to connect, when I can’t change gear by using “+” “-“ on the screen. I am using Zwift Cog and not a classic Gear

**To Reproduce**
1. Enable “Zwift Play”, “Gear” and adjust “Gear gain”
2. Quit the app
3. Connect my Zwift Play to my Ipad via Bluetooth 
4. Restart QZ

**Expected behavior**
I expected to have my Zwift Play connected to QZ. I only see my Home Trainer “found” on the dashboard. 

<img width="380" alt="image" src="https://github.com/user-attachments/assets/8497d979-82c2-46ca-b2c6-edf074f6b88f">

**Device**
 - OS: IpadOS
 - Version 17.5.1

[debug-mer__juil__17_22_12_06_2024.log](https://github.com/user-attachments/files/16274483/debug-mer__juil__17_22_12_06_2024.log)

**Comment by cagnulein** [comment]
thanks i will check this tomorrow, 11pm here
have a great day
Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno mer 17 lug 2024 alle ore 22:35 Félix Liburski <
***@***.***> ha scritto:

**Comment by cagnulein** [comment]
hi @felixlbr I checked the log: I can't see the zwift play at all in bluetooth in this log. Are you sure they aren't connected somewhere else and also did you press a button on both before opening QZ to wake them up?
You can also use the free app nrfconnect to see if you can see them. They are called "zwift play"
Let me know

**Comment by cagnulein** [comment]
https://github.com/user-attachments/assets/5946cd71-2681-4b2f-9571-102d8400ceb2

**Comment by felixlbr** [comment]
[debug-lun__juil__22_15_18_07_2024.log](https://github.com/user-attachments/files/16334023/debug-lun__juil__22_15_18_07_2024.log)

**Comment by cagnulein** [comment]
just sent a new version! let me know!
Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno lun 22 lug 2024 alle ore 16:10 cagnulein ***@***.***> ha
scritto:

**Comment by cagnulein** [comment]
@felixlbr any news?

**Comment by felixlbr** [comment]
![IMG_1780](https://github.com/user-attachments/assets/9ba60d7d-1682-4bfd-8367-390fc7a948bd)

I can still not change gear with my zwift play, I have the impression that gear does not affect the Home trainer Resistance

**Comment by cagnulein** [comment]
Nope, its nor the reason. Send me a fresh debug log. Thanks
Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno ven 26 lug 2024 alle ore 09:34 Félix Liburski <
***@***.***> ha scritto:

**Comment by felixlbr** [comment]
I am currently on vacation, I will be able to send it in one week.

**Comment by cagnulein** [comment]
Ok, no problem. Have a great vacation!
Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno ven 26 lug 2024 alle ore 09:46 Félix Liburski <
***@***.***> ha scritto:

**Comment by cagnulein** [comment]
https://github.com/ajchellew/zwiftplay/issues/5

**Comment by Citizenyolo** [comment]
I have the same issue with my new Zwift Ride.  Same way to replicate it as described above.

Zwift Play does seem to connect and the device appears on the Bluetooth list. If I go to Settings / Advanced setting and pull down the Manual Device list it appears as "Zwift Ride". 

However virtual shifting does not work. The only thing it does is when I press the orange brake lever on my LEFT handlebar it changes the virtual shift with a -2 value and if I press it hard it decreases with a value of about -6 or more. It does not feel more difficult or more easy in watt terms. If I press the same button on the RIGHT side, nothing happens, no values change.

I would also consider assigning a different button for gear shifting. Did you know that the orange brake lever (the long button) has a dual function in Zwift? If you press it softly it is used for steering to that direction. If you press it hard it is used for braking (front or rear).

In Zwift the shifting is assigned to the buttons on the SIDE. You can find two shifters (like on a real roadbike) and in Zwift either of the 2 you press on the right side will shift to smaller virtual cog (more effort) and when you press either on the left side will shift to larger virtual cog (less effort). Maybe you could consider keeping consistency with this approach.

[debug-Tue_Aug_6_18_44_36_2024.log](https://github.com/user-attachments/files/16513401/debug-Tue_Aug_6_18_44_36_2024.log)

I have attached the log file.

**Comment by cagnulein** [comment]
Hi @Citizenyolo thanks for the log.
Checked and understood the issue: your zwift ride is different from the previous one because it sends both controller frome one bluetooth device.
I just handled it and I'm sending you a new build. Check the gear tile in qz to see that if it's changing: in the log that you sent me the gears just moves only negative (from 0 to -99), did you notice this?

About the buttons: yes we can map also other buttons, for first of all, let's see if this it's working, then we can proceed mapping other buttons.

Let me know

**Comment by Citizenyolo** [comment]
Thank You!

Yes, The Virtual gear only progressed in negative increments and only on the left side button (no positive gear change when pressing on the right side button). Also it should probably progress in an increment of 1 at a time and not 2, 4 or 8. Also It changed nothing on the pedalling difficulty in any way.

Shall I go to the Testflight app on my iPhone and install QZ again or I need to wait for it to offer update?

**Comment by cagnulein** [comment]
you will receive a new testflight version by email. for the 2 4 or
increment you can adjust the gears gain to 0.5. then i can also add a new
setting to increase it only to the rising edge of the button. i did it in
this way so that if you want to increase or decrease quickly, you are able
to.

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno mar 6 ago 2024 alle ore 20:33 Zoltan Haas <
***@***.***> ha scritto:

**Comment by Citizenyolo** [comment]
Now it's shifting up and down but there is no change in resistance at all. I can shift from -50 to +50 and have the same watt output at the same rpm.

**Comment by Citizenyolo** [comment]
Another issue is that when I launch IndieVelo both apps (1) IndieVelo and (2) QZ connect to my trainer but when IndieVelo is running the virtual shifting no longer works on QZ. I trie
[debug-Tue_Aug_6_20_52_04_2024.log](https://github.com/user-attachments/files/16514609/debug-Tue_Aug_6_20_52_04_2024.log)
d to shift from the app and the values go up and down but does not change the resistance feel.

**Comment by cagnulein** [comment]
great!
the problem about resistance is that indievelo is not sending resistance
control at all!
did you try to restart indievelo?
did you select resistance control on indievelo?

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno mar 6 ago 2024 alle ore 20:56 Zoltan Haas <
***@***.***> ha scritto:

**Comment by Citizenyolo** [comment]
OK, I restarted IndieVelo and the Zwift Play shifting is moving the "Gears" tile up or down. BUT it does nothing. I can be in gear 500 and pull 54 watt and do the same in -500 gear. There does not seem to be any connection between resistance feel and the gear set with or without IndiVelo is the same.

**Comment by cagnulein** [comment]
> I’ll try try it now but the resistance does not change even if i change virtual gear only on QZ. Zwift play works, chamges gear, wattage is rhe same on QZ zraining

that's normal, i'm applying gears mod only when i receiving a resistance request from any 3rd party apps. of course i can do this as well, but let's focus on indievelo first. I will check the log again, in the meantime, try the same setup with zwift, just to double check that it's an indievelo setup issue.

let me know

**Comment by cagnulein** [comment]
i was checking the log and just a stupid question but it worth to ask: are you connecting the iphone to indievelo right? you don't have to connect your trainer to indievelo

**Comment by Citizenyolo** [comment]
So the set up is as follow (and maybe wrong, but I'm new to this gear :) ) (1) Indivelo is run on an iPad Pro and connects to the trainer AND 
![IMG_0057](https://github.com/user-attachments/assets/7fb4669f-1e5e-4e0f-a2e0-c61bbd493b15)

![IMG_0056](https://github.com/user-attachments/assets/b41618fe-d5bc-4f8b-a6ea-aca01df63290)

![IMG_0055](https://github.com/user-attachments/assets/a6071df7-9bf0-458b-a3a3-7fd12f4c56f3)
(2) on my iPhone 14 Pro Max I have QZ and it also connects to the Wahoo trainer.

**Comment by cagnulein** [comment]
ok that's the issue, on indievelo you have to select the iphone not the kickr core. don't you see the qz device there?

**Comment by Citizenyolo** [comment]
OK, I set it up this way, I can see QZ as a separate trainer in IndieVelo and now everything is working. Changing gears and changing resistance in IndieVelo. The only observation is that there is a 5 to 10 second lag between shifting gear and receiving the resistance change feel.

**Comment by cagnulein** [comment]
![image](https://github.com/user-attachments/assets/21139b33-8196-421e-94ad-ff9dd5048737)

You have to setup this way.

For the lag send me the debug log, I guess I know what is it and it's fixable of course

**Comment by Citizenyolo** [comment]
herre is the log

**Comment by cagnulein** [comment]
@Citizenyolo the log is corrupted, you have to wait the finish of the upload before clicking on the comment button

**Comment by Citizenyolo** [comment]
Here it is again. 
[debug-Tue_Aug_6_21_36_44_2024.log](https://github.com/user-attachments/files/16515085/debug-Tue_Aug_6_21_36_44_2024.log)

**Comment by cagnulein** [comment]
got the issue! i'm sending a new on in 15 minutes!

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno mar 6 ago 2024 alle ore 21:56 Zoltan Haas <
***@***.***> ha scritto:

**Comment by cagnulein** [comment]
yes i have already handled these button for the classic zwift play devices. i just need a debug log from you where you are pressing only the new buttons so i can handle this as well. note: ONLY the new buttons. I'm uploading now the version with the lag fix.
Don't worry for the time, i have my daugther and my wife to my parents in law so i'm a free bird :D

**Comment by Citizenyolo** [comment]
OK :) so, I just start IndieVelo and press say 5-10 times the right 2 buttons and than do the same 5-10 times the left side buttons and you can identify the buttons from the log file?

**Comment by Citizenyolo** [comment]
I can also press DIFFERENT times the 4 buttons (2 on each side) so you can clearly identify them?

**Comment by cagnulein** [comment]
I guess you forgot to add the log :)

**Comment by cagnulein** [comment]
or you are saying that I should look the last one?

**Comment by Citizenyolo** [comment]
No I have not done it, just wanted to ask if that is how I can help? If yes, I'll go and do that and send the log file.

**Comment by cagnulein** [comment]
write the order of you presses, so i can identify them

**Comment by Citizenyolo** [comment]
OK, here is what I did:
Right Top side button: 3x
Right Bottom side button: 6x
Left Top side button: 5x
Left Bottom side button: 10X
[debug-Tue_Aug_6_22_12_54_2024.log](https://github.com/user-attachments/files/16515300/debug-Tue_Aug_6_22_12_54_2024.log)

**Comment by cagnulein** [comment]
perfect, so just to be sure: i'm adding the 2 right buttons to the plus gear and the other 2 left ones to the minus one, right?

**Comment by Citizenyolo** [comment]
Give me 2 mins to double check on Zwift...

**Comment by Citizenyolo** [comment]
Correct: 2 right buttons to the plus gear (higher resistance) and the other 2 left ones to the minus one (lower resistance)

**Comment by cagnulein** [comment]
done! i'm sending you also this one!

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno mar 6 ago 2024 alle ore 22:24 Zoltan Haas <
***@***.***> ha scritto:

**Comment by Citizenyolo** [comment]
Buttons are perfect, same as in Zwift. Lag is much better but it's still there but not 5-10 sec more like 2-4 sec lag.

[debug-Tue_Aug_6_22_41_34_2024.log](https://github.com/user-attachments/files/16515574/debug-Tue_Aug_6_22_41_34_2024.log)

**Comment by cagnulein** [comment]
got the issue, i don't know why but it wasn't fixed the previous one. let me check. perfect for the buttons!

**Comment by Citizenyolo** [comment]
Lag is still there, I did not feel any change vs previous version. Log is attached. This week I'm working from home so you can reach me any time for testing new versions. Thanks!

[debug-Tue_Aug_6_23_35_11_2024.log](https://github.com/user-attachments/files/16515974/debug-Tue_Aug_6_23_35_11_2024.log)

**Comment by cagnulein** [comment]
ok i have to add some other debug log to understand it. tomorrow we will solve it for sure :)
have a good night!

**Comment by Citizenyolo** [comment]
I tested it but something got worse. Now the resistance does not change to any direction, always at about 75 watt. Attached the log.

[debug-Wed_Aug_7_10_01_40_2024.log](https://github.com/user-attachments/files/16520283/debug-Wed_Aug_7_10_01_40_2024.log)

**Comment by cagnulein** [comment]
i don-t see a the connection from indievelo, that's the issue. did you pair as before?

**Comment by Citizenyolo** [comment]
I tried again and I think I connected correctly this time. Before the update the name of the trainer was "QZ" and now it changed to "iPhone 14". That's why I overlooked it. I also attached my heart rate monitor FYI. 

The update does not seem to address the lag, which is about 3-4 sec.

[debug-Wed_Aug_7_10_14_38_2024.log](https://github.com/user-attachments/files/16520500/debug-Wed_Aug_7_10_14_38_2024.log)

Attached the log.

**Comment by cagnulein** [comment]
ok now i see everything and everything is under 200ms. I'm wondering if the issue is that your trainer doesn't get small difference from the gears. try to increase the gears gain for example to 5, so this means that every time you press it it will increase by 5.

**Comment by Citizenyolo** [comment]
> ok now i see everything and everything is under 200ms. I'm wondering if the issue is that your trainer doesn't get small difference from the gears. try to increase the gears gain for example to 5, so this means that every time you press it it will increase by 5.

Where can I set it? In Advanced settings I can see Watt gain, cadence gain, speed gain but no gear gain setting...

**Comment by cagnulein** [comment]
bike options

**Comment by Citizenyolo** [comment]
I set gear gain from 1 to 5. When I change gear it now goes in about 10 or 20 increments. It now feels much more responsive but the change from one gear to the next next is dramatic eg. from 75 watt to 180 watt at same rpm.

Zwift publishes the gear ratio for each gear. you can see the increments is about 0.2 from each gear. The current version feels more like an increment of 1.0 or more per gear change.

![Zwift gears](https://github.com/user-attachments/assets/ccbf749d-18e1-4aa3-a6dd-03ea255e2e7a)

[debug-Wed_Aug_7_10_31_26_2024.log](https://github.com/user-attachments/files/16520706/debug-Wed_Aug_7_10_31_26_2024.log)

**Comment by Citizenyolo** [comment]
If it's any help I can go through the gears in zwift at a constant rpm and tell you for each virtual gear the wattage. May help to calibrate a more consistent feel? Or maybe more settings in QZ to address it?

**Comment by cagnulein** [comment]
Ok so let me summarize it: so now with the gears to 5 you feel that it's changing immediately right? so this means that your trainer has low sensitivity to the inclination changes.
I mean, QZ is applying an offset to the inclination sent from indievelo. Assume that you are at 0% in indievelo, and you have gear 1 in QZ, QZ is sending 0.5% to your trainer.

What you are telling me, is that 0.5% or 1% is not noticable, instead a 2.5% is noticeable, right?

another test that you can do, without QZ, pairing indievelo directly to your trainer, is to see on a hill how much you feel the inclination, i mean if it's the indivelo grade changes from 0.5 to 1 and you don't feel it, that's a confirmation of what i'm saying

If it's so, you just need to find a confortable gears gain, for example, maybe 2 is enough for you

**Comment by Citizenyolo** [comment]
So, a few test:

I tested IndieVelo on my trainer and the gradient changes were between 0% and +/- 1%. The watt meter showed about a 30W change immediately and there was an immediate subtle change in resistance feel. Same is for Zwift natively.

I also fired up Mywhoosh with Belgium (0% gradient all the way) hooked up to QZ. I lowered the gear gain to 2. When I change 1 gear the difference in wattage is about 100 watts. That's a lot for one gear.

On IndieVelo maybe there is a workaround to this, as it does offer to change virtual gear on the screen or with a bluetooth keyboard's +/-. You already successfully reverse engineered the buttons on the Zwift Play so maybe it would be not only easier to link the buttons to Indivelo's +/- bluetooth reciever but also keep Indivelo's famous accuracy intact. Someone seems to have experimented with this idea already: 

https://github.com/jat255/zwift_click_handling

[debug-Wed_Aug_7_11_01_41_2024.log](https://github.com/user-attachments/files/16521112/debug-Wed_Aug_7_11_01_41_2024.log)

**Comment by cagnulein** [comment]
ok so for what you are saying a gears gain of 0.5 or even less it should be the perfection.
Maybe the issue is that qz is sending too much frames on the trainer? 
Try to set the 0.5 gears gain and do just one gear change at a time. do you feel the difference?
It could be a matter of too much frames because usually indievelo or zwift sends the inclination only when it changes, qz, if you are changing the gears continuosly it sends it each time. (so in a second i can send 5 different settings to the trainer max.).

We just need to figure out the limit of your trainer, check the guy here https://github.com/cagnulein/qdomyos-zwift/issues/2470 it has your same setup and it seems ok to him after the lag patch. I'm mentioning him here to discuss about this @jontnswift

**Comment by cagnulein** [comment]
> I may have a quick fix idea for IndiVelo for racers / quicker shifting. You can assign maybe the lower right side buttons on zwift play for transmitting "+" signal for and lower left to transmit "-" signal. You can keep the upper side buttons as is for QZ gear change, it is a duplicate anyway as per Zwift assignment.

sure, do you think this could solve your issue? 
I didn't get from what you are saying if it's a delay thing of qz or a delay thing of your trainer or just a gear gain issue. From your phrase it seems that you need a more precise settings, and yes I can do easily this mod. Let me know if you want to proceed in this way!

**Comment by Citizenyolo** [comment]
That would be great if we can have that +/- set up to the lower buttons. IndieVelo does change gears a bit faster but it is clumsy to use the touchscreen for that.

**Comment by cagnulein** [comment]
hah no wait i misread the message: i guess you would like to set a 2x gear with the top button. No i can't emulate a keyboard button to indievelo. I don't understand how it could be faster on indievelo: it's the same message sent from a device to another. How much faster is it? in the order of 0.5s?

**Comment by Citizenyolo** [comment]
I may be completely wrong, I'm no expert, just brainstorming :) . But in that scenario I would choose the Wahoo trainer in IndieVelo (not QZ). QZ as an app or "virtual trainer" connected to my iPad (running Indivelo app) could send +/- signal whenever I press the correct keys in Zwift Play. This would trigger Indivelo's built in virtual gear change which constantly listens to +/- input for that from a bluetooth keyboard.

**Comment by cagnulein** [comment]
QZ can't broadcast itself like a bluetooth keyboard, that's the issue. Indievelop app should have some APIs to do this in order to implement this. The only way to do what you are asking is using indievelo on PC and play a macro to use + and - minus button on the PC itself. Ipad and ios devices are out of the game.

But again I would like to understand where the delays comes in: i mean from my point of view, qz is a man on the middle, i accept that there is a delay because there is a one more device in, but i guess it's in the order of some hundreds millseconds, not more. If there is more, there is somethign else that we have to understand (unfortunately it's not easy  on remote). For example do you see the gears change immediately on qz? If so we can say that the zwift play controller to QZ is fine, and we can focus on the trainer to qz communication only.

**Comment by Citizenyolo** [comment]
I think you are right. Gears now DO change almost immediately on QZ. I think the delay is about 1/5th of a second only.

I also did another test, which I should have done earlier. I compared the speed of watt change by changing gear on Indivelo's on screen touch enabled virtual gear buttons. If I shift up  quickly 5 gears it also takes a few seconds for the full impact to arrive. Your solution now delivers about 99% the same experience. The only thing I can still finetune is the gear change, which I think will be between 0.3 and 0.5 to replicate the zwift gear ratio steps.

I would leave the buttons as is, users coming from Zwift would have the same experience.

Overall I think we can close this: Questo lavoro è perfetto.

Thank You!!
